### ISSUE TITLE
Loss of child proxy

### ISSUE THREAD

**Issue by ordinaryorange** [mfidemraizer/trackerdog#11]
Came across another one today.
The issue arises with the use of ToUntrackedEnumerable
Here is the test code that repo's the issue
``` c#
void Main()
{
	TrackerDog.Configuration.TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<A>(), Track.ThisType<B>(), Track.ThisType<C>());  
	var a = new A();
	var b = new B();
	var c = new C();
	a.B = b;
	b.CSet = new HashSet<C>();
	b.CSet.Add(new C() { MyProperty=10 });
	b.CSet.Add(new C() { MyProperty=20 });
    var aa = a.AsTrackable();
	aa.B.CSet.Dump(); //CSet is a proxy
	var listOfTrackables = new List<A>();
	listOfTrackables.Add(aa);
	var n = listOfTrackables.ToUntrackedEnumerable(typeof(IList<>));
	aa.B.CSet.Dump(); //CSet is no longer a proxy. It is just a plain HashSet<>
}

public class A 
{
	public virtual B B { get; set; }
}

public class B
{
	
	public virtual ISet<C> CSet { get; set; }
}

public class C 
{
	public virtual int MyProperty { get; set; }
}


```
Can you confirm your end ?

**Comment by ordinaryorange** [comment]
Bit more testing, get the same result if I use the `ToUntracked()` extension method

**Comment by mfidemraizer** [comment]
Now I won't be able to throw a definitive answer, but I believe that it's the expected behavior.

`ToUntrackedEnumerable()` is recursive.

Are you looking for being able to tell this method and its brother `ToUntracked()` that you don't want to **recursively** turn a given collection/object to untracked?

**Comment by ordinaryorange** [comment]
No that't not the issue. If you recheck the code the following is happening.
1. I Create a object graph of POCO's (root is object `a`)
2. I then turn the object graph into a proxy graph. (so the root and all decendants are now all trackables) (object representing this is `aa`)
3. I then ask the `aa` instance for the original POCO that it wraps around (assuming I did not know what it was in the first place) (effectively I want object `a` back)
3. I do get `a` back as expected and it is a POCO graph, but if I go an inspect `aa` I notice that the root is still a proxy, but decentants have changed back to POCO's. They should still be proxies !

Do you use linpad ? It is thes best $ I have ever spent. You could copy the above code and past it in (after adding the trckerdog references) and you can see the result faster than it would take to ope studio. I use it for running test code all the time, and to nut out class behaviours etc.

**Comment by mfidemraizer** [comment]
@ordinaryorange  If I'm not mistaken, you might be misunderstanding what `ToUntracked()` does. It doesn't give you the object without modifying the graph but it turns the graph into a non-trackable object.

Maybe I can implement a new method called `Extract()` which can give you the same result without modifying the graph (it'll be just getting `CurrentValue` of each property tracking).

Is this what you really want?

**Comment by mfidemraizer** [comment]
I've double-checked the problem and you're right... I'll investigate it further.

**Comment by mfidemraizer** [comment]
@ordinaryorange I'm not sure if it's a DynamicProxy bug or a misunderstanding of how it works.

Actually, when an instance is untracked, I do something like this:

`var target = ((IProxyTargetAccessor)trackable).DynProxyGetTarget();
`

The whole `target` should be the aggregate root's proxy target (i.e. the object being wrapped...). 

For now, I don't understand why getting a given property using reflecting and setting it giving `target`  as the property's instance sets the wrapped property too...

 `target.GetProperty("Whatever").SetValue(target, null);` 

I'll continue to investigate what's going on...

**Comment by mfidemraizer** [comment]
@ordinaryorange  I've published a Q&A on StackOverflow: http://stackoverflow.com/questions/38640557/proxy-target-property-being-set-to-proxy-property-counterpart

**Comment by mfidemraizer** [comment]
Finally I believe that the issue behavior is the expected one. If you set the target property, the proxy object changes because it wraps the whole target object.

I believe that a definitive solution to this issue is that `ToUntracked()` should return a _deep clone_ of the target, instead of the target itself. I'll work on it, I believe that it should be as easy as cloning the object with recursion...

**Comment by mfidemraizer** [comment]
@ordinaryorange Hey,

Finally I managed to solve it performing a deep clone of the whole proxy target (the underlying object that's being wrapped by the whole proxy...).

For now, it's not the most performant solution but it works fine. I use the serialization/deserialization approach. There're more powerful ones like ones using expression trees (and there're many libraries out there implementing this approach) but I found that they would need extra configuration.

I believe that for the TrackerDog case it should work fine and if you or any other developer finds performance issues, everyone can fill a new issue to optimize the deep clone approach.

Anyway, this fix will be available along with first *_2.0_ release in few days, where I've changed the configuration model and many other things. See #12 for more details ;)

Oh, and I forgot to credit [David L](http://stackoverflow.com/users/1165998/david-l) from StackOverflow for [this answer](http://stackoverflow.com/questions/38745047/declared-properties-wont-be-serialized-if-its-declaring-type-is-a-dynamic-objec/38745115#38745115) ;)
